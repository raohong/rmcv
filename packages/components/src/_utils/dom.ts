export const isWindow = (win: any): win is Window => win === window;

const getDocRect = () => {
  const html = document.documentElement;

  return {
    left: 0,
    top: 0,
    bottom: html.clientHeight,
    right: html.clientWidth,
    width: html.clientWidth,
    height: html.clientHeight,
  };
};

export type IBCR = {
  left: number;
  top: number;
  bottom: number;
  right: number;
  width: number;
  height: number;
};

export const getScrollOffset = (elem: Window | Element) => {
  if (isWindow(elem)) {
    return {
      scrollLeft: elem.pageXOffset,
      scrollTop: elem.pageYOffset,
    };
  }

  return {
    scrollLeft: elem.scrollLeft,
    scrollTop: elem.scrollTop,
  };
};

export const getScrollSize = (elem: Window | Element) => {
  if (isWindow(elem)) {
    return {
      scrollWidth: document.documentElement.scrollWidth,
      scrollHeight: document.documentElement.scrollHeight,
    };
  }

  return {
    scrollWidth: elem.scrollWidth,
    scrollHeight: elem.scrollHeight,
  };
};

export const getBoundingClientRect = (elem: Window | Element) => {
  if (isWindow(elem)) {
    return getDocRect();
  }

  const rect = elem.getBoundingClientRect();

  return {
    left: rect.left,
    top: rect.top,
    bottom: rect.bottom,
    right: rect.right,
    height: rect.height,
    width: rect.width,
  };
};
