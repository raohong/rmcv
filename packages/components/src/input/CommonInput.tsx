import { useControllableValue } from '@rmc-vant/hooks';
import { ClearFilled } from '@rmc-vant/icons';
import { isEmpty, isNumber, omit } from '@rmc-vant/utils';
import classNames from 'classnames';
import React, { useState } from 'react';
import { useConfigContext } from '../config-provider';
import Touchable from '../touchable';
import type { CommonInputProps } from './interface';

const CommonInput = (
  {
    className,
    size,
    maxLength,
    placeholder,
    border,
    readonly,
    disabled,
    formatter,
    onBlur,
    onFocus,
    clearIcon,
    clearTrigger,
    clearable,
    suffix,
    addonAfter,
    addonBefore,
    prefix,
    onClear,
    showWorldLimit,
    onInput,
    inputAlign,
    wrapperProps,
    inputRef,
    component = 'input',
    formatTrigger = 'onChange',
    ...props
  }: CommonInputProps,
  ref: React.Ref<any>,
) => {
  const { getPrefixCls } = useConfigContext();
  const [value, setValue] = useControllableValue<string | undefined>(props);
  const [focused, setFocused] = useState(false);

  const internalMaxLength =
    isNumber(maxLength) && maxLength >= 0 ? maxLength : Infinity;

  const handleFocus = (
    evt: React.FocusEvent<HTMLInputElement | HTMLTextAreaElement>,
  ) => {
    (onFocus as React.FormEventHandler<HTMLInputElement | HTMLTextAreaElement>)?.(
      evt,
    );
    setFocused(true);
  };

  const handleBlur = (
    evt: React.FocusEvent<HTMLInputElement> | React.FocusEvent<HTMLTextAreaElement>,
  ) => {
    (onBlur as React.FormEventHandler<HTMLInputElement | HTMLTextAreaElement>)?.(
      evt,
    );
    setFocused(false);

    if (formatTrigger === 'onBlur' && formatter && value) {
      setValue(formatter(value)?.slice(0, internalMaxLength));
    }
  };

  const handleChange = (
    evt: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>,
  ) => {
    let nextValue = evt.target.value;

    if (formatTrigger === 'onChange' && formatter) {
      nextValue = formatter(nextValue);
    }

    setValue(nextValue?.slice(0, internalMaxLength));
  };

  const handleClear = () => {
    setValue('');
    onClear?.();
  };

  const cls = getPrefixCls('input');

  const clearIconVisible =
    clearable && (clearTrigger === 'focus' ? focused : true) && !!value;
  const counterVisible = isNumber(maxLength) && maxLength >= 0 && showWorldLimit;
  const Com = component;

  return (
    <div
      ref={ref}
      {...wrapperProps}
      className={classNames(`${cls}-wrapper`, wrapperProps?.className)}
    >
      <div className={`${cls}-container`}>
        {!isEmpty(addonBefore) && (
          <span className={`${cls}-addon-before`}>{addonBefore}</span>
        )}
        {!isEmpty(prefix) && <span className={`${cls}-prefix`}>{prefix}</span>}
        <Com
          onBlur={handleBlur}
          onChange={handleChange}
          onFocus={handleFocus}
          placeholder={placeholder}
          maxLength={internalMaxLength}
          value={value ?? ''}
          className={classNames(
            cls,
            {
              [`${cls}-disabled`]: disabled,
              [`${cls}-focused`]: focused,
              [`${cls}-readonly`]: readonly,
              [`${cls}-align-${inputAlign}`]: inputAlign && inputAlign !== 'left',
            },
            className,
          )}
          disabled={disabled}
          readOnly={readonly}
          ref={inputRef as unknown as React.Ref<any>}
          {...omit(props as any, ['children', 'onChange', 'defaultValue', 'value'])}
        />

        {clearIconVisible && (
          <Touchable onClick={handleClear} component="button">
            {isEmpty(clearIcon) ? (
              <ClearFilled className={`${cls}-clear-icon`} />
            ) : (
              clearIcon
            )}
          </Touchable>
        )}
        {!isEmpty(suffix) && <span className={`${cls}-suffix`}>{suffix}</span>}
        {counterVisible && component === 'input' && (
          <span className={classNames(`${cls}-counter`)}>
            {(value || '').length} / {maxLength}
          </span>
        )}
        {!isEmpty(addonAfter) && (
          <span className={`${cls}-addon-after`}>{addonAfter}</span>
        )}
      </div>
      {counterVisible && component === 'textarea' && (
        <span className={classNames(`${cls}-counter`, `${cls}-textarea-counter`)}>
          {(value || '').length} / {maxLength}
        </span>
      )}
    </div>
  );
};

export default React.forwardRef(CommonInput);
