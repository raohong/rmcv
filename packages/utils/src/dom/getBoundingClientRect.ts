import { isElement } from './instanceOf';
import isWindow from './isWindow';

export interface IBCR {
  left: number;
  top: number;
  bottom: number;
  right: number;
  width: number;
  height: number;
}

const getClientRect = (node: Element): IBCR => {
  const rect = node.getBoundingClientRect();

  return {
    left: rect.left,
    top: rect.top,
    right: rect.right,
    bottom: rect.bottom,
    width: rect.width,
    height: rect.height,
  };
};

const getWinClientRect = (): IBCR => {
  const w = window.innerWidth;
  const h = window.innerHeight;

  return {
    left: 0,
    top: 0,
    right: w,
    bottom: h,
    width: w,
    height: h,
  };
};

export const getBoundingClientRect = (node?: null | Window | Element) => {
  if (node === null || node === undefined || isWindow(node)) {
    return getWinClientRect();
  }

  return isElement(node) ? getClientRect(node) : getWinClientRect();
};
