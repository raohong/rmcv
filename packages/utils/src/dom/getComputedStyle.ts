import getWindow from './getWindow';

const getComputedStyle = (node: Element) => getWindow(node).getComputedStyle(node);

export default getComputedStyle;
