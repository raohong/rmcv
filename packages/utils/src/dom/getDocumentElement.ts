import { isElement } from './instanceOf';

export const getDocumentElement = (node?: Node | Window | null) =>
  (isElement(node) ? node.ownerDocument : window.document).documentElement;

export default getDocumentElement;
