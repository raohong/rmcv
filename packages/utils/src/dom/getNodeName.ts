const getNodeName = (node: Window | Node) =>
  // @ts-ignore
  (node.nodeName ?? '').toLowerCase?.();

export default getNodeName;
