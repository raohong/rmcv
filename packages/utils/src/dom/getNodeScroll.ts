import getWindow from './getWindow';
import { isHTMLElement } from './instanceOf';

const getHTMLElementScroll = (node: HTMLElement) => ({
  scrollLeft: node.scrollLeft,
  scrollTop: node.scrollTop,
});

const getWindowScroll = (node: Node | Window) => {
  const win = getWindow(node);

  return {
    scrollLeft: win.scrollX || win.pageXOffset,
    scrollTop: win.scrollY || win.pageYOffset,
  };
};

const getNodeScroll = (node: Node | Window) => {
  if (node === getWindow(node) || !isHTMLElement(node)) {
    return getWindowScroll(node);
  }

  return getHTMLElementScroll(node);
};

export default getNodeScroll;
