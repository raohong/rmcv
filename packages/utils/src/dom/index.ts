export type { IBCR } from './getBoundingClientRect';

export { default as getNodeScroll } from './getNodeScroll';

export { default as getScrollParent } from './getScrollParent';

export { default as getBoundingClientRect } from './getBoundingClientRect';

export { default as getNodeScrollSize } from './getNodeScrollSize';

export { default as listenScrollParents } from './listenScrollParents';

export { default as getDocumentElement } from './getDocumentElement';

export * from './instanceOf';
