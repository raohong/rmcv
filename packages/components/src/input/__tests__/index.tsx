import { fireEvent } from '@testing-library/dom';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import CommonInput from '../CommonInput';

const testId = 'input';

test('snapshot', () => {
  const com = render(<CommonInput />);

  expect(com.asFragment()).toMatchSnapshot();
});

test('render with value', () => {
  render(<CommonInput value="input" data-testid={testId} />);

  expect(screen.getByDisplayValue('input')).toBeInTheDocument();
});

test('render with onChange', () => {
  const onChange = jest.fn();

  render(<CommonInput onChange={onChange} data-testid={testId} />);

  userEvent.type(screen.getByTestId(testId), '123');

  expect(onChange).toHaveBeenCalledWith('123');
});

test('render with formatter', () => {
  const formatter = jest.fn((val) => val.replace(/\D/g, ''));
  const com = render(<CommonInput formatter={formatter} data-testid={testId} />);

  userEvent.type(screen.getByTestId(testId), '12k');

  expect(formatter).toHaveBeenCalledWith('12k');
  expect(screen.getByDisplayValue('12')).toBeInTheDocument();

  com.rerender(
    <CommonInput
      formatTrigger="onBlur"
      formatter={formatter}
      data-testid={testId}
    />,
  );

  userEvent.type(screen.getByTestId(testId), 'blur');
  expect(screen.getByDisplayValue('12blur')).toBeInTheDocument();

  fireEvent.focusOut(screen.getByTestId(testId));
  expect(formatter).toHaveBeenCalledWith('12blur');
  expect(screen.getByDisplayValue('12')).toBeInTheDocument();
});
