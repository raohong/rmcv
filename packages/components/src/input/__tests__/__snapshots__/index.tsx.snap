// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<DocumentFragment>
  <div
    class="rmcv-input-wrapper"
  >
    <div
      class="rmcv-input-container"
    >
      <input
        class="rmcv-input"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;
